src/main/java/com/iota/iri/utils/dag/DAGUtils.java

```diff
@@ -0,0 +1,366 @@
+     */
+    public static DAGUtils get(Tangle tangle) {
+        DAGUtils instance = instances.get(tangle);
+        if(instance == null) {
+            instances.put(tangle, instance = new DAGUtils(tangle));
+        }
+
+        return instance;
+    }
+
+    /**
+     * Constructor of the class which allows to provide the required dependencies for the implemented methods.
+     *
+     * Since the constructor is protected we are not able to manually create instances of this class and have to
+     * retrieve them through their singleton accessor {@link #get}.
+     *
+     * @param tangle database interface used to retrieve data
+     */
+    protected DAGUtils(Tangle tangle) {
+        this.tangle = tangle;
+    }
+
+    //region TRAVERSE APPROVERS (BOTTOM -> TOP) ////////////////////////////////////////////////////////////////////////
+
+    /**
+     * This method provides a generic interface for traversing all approvers of a transaction up to a given point.
+     *
+     * To make the use of this method as easy as possible we provide a broad range of possible parameter signatures, so
+     * we do not have to manually convert between milestones, transactions and hashes (see other methods with the same
+     * name).
+     *
+     * It uses an non-recursive iterative algorithm that is able to handle huge chunks of the tangle without running out
+     * of memory. It creates a queue of transactions that are being examined and processes them one by one. As new
+     * approvers are found, they will be added to the queue and processed accordingly.
+     *
+     * Every found transaction is passed into the provided condition lambda, to determine if it still belongs to the
+     * desired set of transactions and only then will be passed on to the currentTransactionConsumer lambda.
+     *
+     * @param startingTransaction the starting point of the traversal (does not get passed to the consumer)
+     * @param condition a lambda function that allows to control how long the traversal should continue
+     * @param currentTransactionConsumer a lambda function that allows us to "process" the found transactions
+     * @param processedTransactions a set of hashes that shall be considered as "processed" already and that will
+     *                              consequently be ignored in the traversal
+     * @throws TraversalException if anything goes wrong while traversing the graph and processing the transactions
+     */
+    public void traverseApprovers(TransactionViewModel startingTransaction,
+                                  TraversalCondition condition,
+                                  TraversalConsumer currentTransactionConsumer,
+                                  Set<Hash> processedTransactions) throws TraversalException {
+        final Queue<TransactionViewModel> transactionsToExamine = new LinkedList<>();
+        try {
+            startingTransaction.getApprovers(tangle).getHashes().stream().forEach(approverHash -> {
```

Comment:
But getApprovers does exactly that?

```
    public ApproveeViewModel getApprovers(Tangle tangle) throws Exception {
        if(approovers == null) {
            approovers = ApproveeViewModel.load(tangle, hash);
        }
        return approovers;
    }
```

This comment would only make sense if we would not have the tvm's already but the hashes, because then we could skip fetching the TVM.

In fact since getApprovers caches the result i think the existing way tends to be faster than what you wrote.